### PR TITLE
bitcoin: install desktop file and pixmaps

### PR DESCRIPTION
**Pull request by offlinehacker**
###### Motivation for this change

Add desktop file and pixmaps, so we have better intergration

###### Things done

I had to switch source from one on bitcoin.org to github, as distributed package does not include pixmaps. This is also something that archlinux uses.

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Notify maintainers

cc @roconnor

### CODE REVIEW

File: pkgs/applications/blockchains/bitcoin.nix [hunk 1]
```diff
@@ -1,17 +1,27 @@
...
-           ];
-    sha256 = "5c7d93f15579e37aa2d1dc79e8f5ac675f59045fceddf604ae0f1550eb03bf96";
+  packaging = fetchFromGitHub {
+    owner = "bitcoin-core";
+    repo = "packaging";
+    rev = "a48094dca1113fb6096768993d1b80d1a4ab5871";
+    sha256 = "0kkbpw4kcxffdik35vvf35vdkcpjfacj1m6vxr73v4f1fzw1kx9m";
+  };
+
+  src = fetchFromGitHub {
```
Comment (jb55) [inline pkgs/applications/blockchains/bitcoin.nix]: why was this changed to the github repo? I would prefer if we stuck to the tarballs signed with the bitcoin release key.

**Review by jb55 — CHANGES_REQUESTED** [review]

File: pkgs/applications/blockchains/bitcoin.nix (hunk 1 above)
Comment (offlinehacker) [reply pkgs/applications/blockchains/bitcoin.nix]: I had to switch source from one on bitcoin.org to github, as distributed package does not include pixmaps. This is also something that archlinux uses.

Comment (offlinehacker) [reply pkgs/applications/blockchains/bitcoin.nix]: I can also only copy pixmaps from GitHub and build from existing source

Comment (jb55) [reply pkgs/applications/blockchains/bitcoin.nix]: perhaps it would be better if we added the pixmaps to the packaging repo?

Comment (jb55) [reply pkgs/applications/blockchains/bitcoin.nix]: or fetch it directly from github ? https://github.com/bitcoin-core/packaging/pull/27/files#diff-5fde7a6d86053f0e1d88c0a2a238941f

Comment (offlinehacker) [reply pkgs/applications/blockchains/bitcoin.nix]: Agreed, i also made downloads optional only if `withGui` is true, so it should only affect `bitcoin` package and not `bitcoind`

**Review by jb55 — APPROVED** [review]
nice! ACK 8d6d7b8ad6593c213bc29bf1d5d4bcc47c49dbe9

File: pkgs/applications/blockchains/bitcoin.nix [hunk 2]
```diff
@@ -14,6 +18,16 @@ stdenv.mkDerivation rec{
     sha256 = "5c7d93f15579e37aa2d1dc79e8f5ac675f59045fceddf604ae0f1550eb03bf96";
   };
 
+  desktop = optional withGui (fetchurl {
+    url = "https://raw.githubusercontent.com/bitcoin-core/packaging/${majorVersion}/debian/bitcoin-qt.desktop";
+    sha256 = "0cpna0nxcd1dw3nnzli36nf9zj28d2g9jf5y0zl9j18lvanvniha";
+  });
+
+  pixmap = optional withGui (fetchurl {
+    url = "https://raw.githubusercontent.com/bitcoin/bitcoin/v${version}/src/qt/res/src/bitcoin.svg";
```
Comment (prusnak) [inline pkgs/applications/blockchains/bitcoin.nix]: It would be better to use

```
url = "https://raw.githubusercontent.com/bitcoin/bitcoin/v${version}/share/pixmaps/bitcoin128.png";
```

**Review by prusnak — CHANGES_REQUESTED** [review]

File: pkgs/applications/blockchains/bitcoin.nix (hunk 2 above)
Comment (jb55) [reply pkgs/applications/blockchains/bitcoin.nix]: > So we don't install an SVG file with the PNG extension.

oops I missed this in review. ideally we would just fix the `install` command to use the proper filename.

File: pkgs/applications/blockchains/bitcoin.nix [hunk 3]
```diff
@@ -14,6 +18,16 @@ stdenv.mkDerivation rec{
     sha256 = "5c7d93f15579e37aa2d1dc79e8f5ac675f59045fceddf604ae0f1550eb03bf96";
   };
 
+  desktop = optional withGui (fetchurl {
+    url = "https://raw.githubusercontent.com/bitcoin-core/packaging/${majorVersion}/debian/bitcoin-qt.desktop";
```
Comment (roconnor) [inline pkgs/applications/blockchains/bitcoin.nix]: I'd prefer you use [`majorMinor`](https://github.com/NixOS/nixpkgs/blob/909904d0ae8750cd6df237b7b439fd2ffeba6273/lib/versions.nix#L45) instead of defining your own `majorVersion`.

**Comment by Lassulus** [comment]
superseded by https://github.com/NixOS/nixpkgs/pull/74741

**Comment by roconnor** [comment]
https://github.com/NixOS/nixpkgs/pull/74741 didn't implement the desktop file and pixmaps, so I'm reopening this PR.  It will need to rebased, but I think we can proceed after that.

File: pkgs/applications/blockchains/bitcoin.nix (hunk 2 above)
Comment (offlinehacker) [reply pkgs/applications/blockchains/bitcoin.nix]: Ok, i fixed url

File: pkgs/applications/blockchains/bitcoin.nix (hunk 3 above)
Comment (offlinehacker) [reply pkgs/applications/blockchains/bitcoin.nix]: Thx, implemented

**Comment by offlinehacker** [comment]
Rebased and implemented suggested changes, should be good to merge

**Comment by roconnor** [comment]
If you are okay with it, I'd prefer if you move the `desktop` and `pixmap` assignments into a `let` expression surrounding the `mkDerivation` (and drop the `optional`).  Then use splices in the `postInstall` as follows:

    postInstall = optional withGui ''
      install -Dm644 ${desktop} $out/share/applications/bitcoin-qt.desktop
      install -Dm644 ${pixmap} $out/share/pixmaps/bitcoin128.png
    '';

While it is a somewhat a matter of preference, I prefer to have lazy evaluation manage what does and does not need to be downloaded.

**Comment by roconnor** [comment]
Ah, you'll need to float the `version` number out into the `let` expression too and `inherit` it.  I think I'm okay with that.

**Comment by offlinehacker** [comment]
> While it is a somewhat a matter of preference, I prefer to have lazy evaluation manage what does and does not need to be downloaded.

Download is lazy already due `optional`

**Comment by offlinehacker** [comment]
> Ah, you'll need to float the version number out into the let expression too and inherit it. I think I'm okay with that.

I do not see how this improves anything, why do you think this is cleaner?

**Comment by offlinehacker** [comment]
I would like to keep things cleaner without using `let` expression, as I do not see benefits here. I do not think we have guidelines about that, so I am not against it, just do not see benefits.

**Comment by roconnor** [comment]
Download isn't controlled by lazy evaluation.  You had to explicitly mark some parts as `optional withGui` and other parts as not.

Very generally speaking, the more we are explicit about this sort of thing, the more likely it is to cause problems in future revisions when things change and we end up unnecessarily downloading objects for build configurations that do not need them.  If we make things implicit than the downloads will exactly occur when the expressions `${desktop}` and `${pixmap}` end up in the derivation structure's value.

However you strongly prefer things the way you've written them, I won't object any further.

**Comment by offlinehacker** [comment]
Ok, after reviewing again I agree with suggested changes, as it removes unnecessary `optional` expression, and also putting `version` into `let` removes unnecessary `rec`. Less code means less room for error, so I will do that, thanks.

**Comment by offlinehacker** [comment]
should be good now

**Comment by roconnor** [comment]
Were you planning on eliminating the `rec`?

**Comment by offlinehacker** [comment]
I couldn't since conditional with `doCheck` uses it here: https://github.com/NixOS/nixpkgs/blob/ad9b68ca0403a804e3acab7b42561e78794cc07a/pkgs/applications/blockchains/bitcoin.nix#L47 Since it's no-op I can remove it or refactor in `let` block, but just keeping `rec` was easiest, what do you prefer?

**Comment by roconnor** [comment]
I see.  I'm good either way, but if you do choose to abstract the `doCheck` then move to a top-level function argument with a default value of `true`.

**Comment by Lassulus** [comment]
whupsie, sorry for closing to early, say when this is ready for merge

**Comment by roconnor** [comment]
I don't know what the correct `-m` permissions ought to be, but `644` seems like a fairly commonly used value.  Looks good to me. @prusnak or @jb55 did you want to re-review?

**Review by jb55 — APPROVED** [review]
LGTM

**Review by roconnor — APPROVED** [review]
